### PR TITLE
Improve map test coverage

### PR DESCRIPTION
**Pull request by abendt**
# Context

See issue https://github.com/arrow-kt/arrow/issues/2894

# Changes

* add more tests for map to improve test coverage
   i have ported some of the tests from here: https://github.com/arrow-kt/arrow-core/tree/master/arrow-core-test/src/main/kotlin/arrow/core/test/laws
* ensure that the generated maps used in the test share enough keys 
   otherwise not all code paths are executed. i have added a test and left some comments that ensures that the generated maps share enough keys. With the previous solution often maps didn't share any keys. I was also thinking about to implement  a specialized Arb for this task (see discussion here https://github.com/kotest/kotest/issues/3459)
       
Edit: with awesome help from the kotest slack i implemented a specialized Arb (see the linked issue)

### CODE REVIEW

File: arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/map.kt [hunk 1]
```diff
@@ -392,7 +392,7 @@ public inline fun <K, reified R> Map<K, *>.filterIsInstance(): Map<K, R> =
  */
 public fun <K, A, B> Map<K, A>.align(b: Map<K, B>): Map<K, Ior<A, B>> =
   (keys + b.keys).mapNotNull { key ->
-    Ior.fromNullables(this[key], b[key])?.let { key to it }
+    Ior.fromNullables(this[key], b[key])!!.let { key to it }
   }.toMap()
```
Comment (serras) [inline arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/map.kt:396]:
```suggestion
  (keys + b.keys).associate { key ->
    key to Ior.fromNullables(this[key], b[key])!!
  }
```

**Review by serras — APPROVED** [review]
Nice work! I've left a small comment, because I think that `associate` reads better than `map` + `toMap` most of the times.

File: arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/map.kt [hunk 2]
```diff
@@ -450,7 +450,7 @@ public fun <K, A> Map<K, Option<A>>.filterOption(): Map<K, A> =
  * Returns a Map containing all elements that are instances of specified type parameter R.
  */
 public inline fun <K, reified R> Map<K, *>.filterIsInstance(): Map<K, R> =
-  mapNotNull { it as? R }
+  mapNotNull { it.value as? R }
```
[the remark this answers is not in the document]
Comment (abendt) [reply arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/map.kt]: ok, will add it and see if i can add a test for that case

[the remark this answers is not in the document]
Comment (abendt) [reply arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/map.kt]: i have added another test for maps with null values. also needed to add a cast to the impl. 
please have another look @serras

[the remark this answers is not in the document]
Comment (serras) [reply arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/map.kt]: Looks OK to me 👍
